### PR TITLE
Migrating grommet to be a monorepo

### PR DESCRIPTION
**Pull request by ShimiSun**
This PR is a POC/suggestion of migrating Grommet repo to be a monorepo.

Migrating to the monorepo is mostly relying on the concept of 'yarn workspaces' and this content: https://yarnpkg.com/lang/en/docs/workspaces/, 
https://www.youtube.com/watch?v=rdeBtjBNcDI&t=

This monorepo presents two modules: 
  1. 'grommet'  
  2. 'templates/starter'. 
(In the future we will like to enhance the structure and add more patterns nested under 'templates', and also add modules for 3. 'grommet-site', 4. 'grommet-icons' etc as needed. )

The current changes are allowing the 'starter' module to run independently, yet coupled to grommet core local changes, and respond to the changes in real-time.
Each template, in this example 'starter', has its own package.json file and can run as an independent application. See its /grommet/templates/starter/README.md for more info.

Note: Running 'yarn install' anywhere in the hierarchy will always install ALL the modules in the workspaces.

All of grommet-core code is moved into 'grommet' directory, with a peer directory of 'templates'. Both are considered as workspaces and managed with the root package.json file.

```
{
	"private": true,
	"workspaces": [
		"grommet",
		"templates/*"
        ]
}
```

To simplify the structure, after running `yarn install` this is what we'll get:
```
--grommet
	--grommet
		package.json
	--templates
		--starter 
			package.json 
	package.json  // defines the workspaces
	node_modules  // only one to include all dependencies
```

Important note, there were no changes made in grommet-core, only on its package.json file to point to the correct main path of  "src/js/index.js", which has all the grommet core exports.

TODOs: 
1. Learn more on how to leverage Lerna when adaption new repos into our monorepos.
2. Make sure we are not losing the git history of moved files - should be good though.
3. After approval of the current monorepo structure, add the rest of the templates to the templates directory.
4. Connect all templates to codesandbox - might require a change with the location of each module's index.html.
5. Figure out the deployment of grommet-site & grommet-icons as modules on the monorepo
6. Understand the changes in the Release process of grommet, and fix Travis
7. Update the package.json paths of grommet-icons, grommet-site etc, to point to the right directory (as long as they are located on a standalone repo).
8. Test all with both 'yarn' and 'npm', currently tested only yarn.
9. Think how to avoid or rename the path of grommet/grommet/grommet on Github
10. Do we need a LICENSE file for every repo?
11. Understand if there is a preference of the port, templates are currently using port 3000, but can be changed easily if need adjustments.
12. Merge all pending PRs before merging the monorepo
13. Get some sleep

### CODE REVIEW

File: grommet/package.json [hunk 1]
```diff
@@ -0,0 +1,148 @@
+{
+  "name": "grommet",
+  "version": "2.4.0",
+  "main": "src/js/index.js",
```
Comment (ShimiSun) [inline grommet/package.json]: This is the only line that was changed in this file.

**Comment by oorestisime** [comment]
Heya,

this is a great start :)

Couple of things to consider maybe:

- are we really needing each starter/pattern/template to be a package and more so to be a workspace? i am not sure if it makes sense to publish such things, in my understanding they are only meant to be example applications and the user can c/p what they want around. In addition each starter could depend on a different number of things so if as you say running yarn installs ALL the modules then it might make them heavier to install.

- if the goal is to have grommet-icons grommet-site etc in it then eventually this would need lerna or so. Does it make sense to structure it directly in a lerna compatible way? it seems to me that lerna searches packages in a specific folder. how it is structured now this won't be possible since grommet seems to be like a top level directory. it would later need to be moved again. which brings me to my next point

- is it necessary to start migrating current structure to monorepo if the only repo there is grommet? the way the starter templates are structured even now don't require any change on grommet itself. they just need to depend on grommet.

**Comment by ShimiSun** [comment]
@oorestisime thanks for the feedback! good questions.

My end-goal was to: 
1. Allow users to actually run the templates independently via yarn/npm/codesandbox. 
2. Use the templates as regression when grommet-core is going through changes.
 
I don't see many purposes for the templates directory in grommet if it doesn't allow it. which lead me to this design of having package.json for each template.

I don't see a problem adding Lerna once we approve this structure, it's definitely on the roadmap if we want to migrate grommet-site and icons, and I don't see a problem adding it in the future. I've yet completed my research or completly tested Lerna, and I'm not sure we need it at the moment since workspaces did the trick on two modules and Lerna might be an overkill.

This PR is only a POC, it will not be merged like that, we will merge it with all the patterns once we agree on the rest of the puzzle peieces. adding the patterns should be easy to do once we approve of it.

**Comment by atanasster** [comment]
Shimi, the goals make a lot of sense, great job.

I agree with separate package.json for each template, so they can be installed independently.

My inclination would be to have two repos though to avoid large size just for grommet core library download
- grommet-core (which could or not include grommet-icons)
- grommet-examples, to include grommet-site, icons-site, templates and all other standalone projects

**Comment by oorestisime** [comment]
Sorry i was probably not clear. i didn't mean not have them as a separate folder with their own package.json but actually as a separate publishable workspace (if lerna comes in place)

@atanasster do you mean a large size when cloning the repo? because the download size on install wouldn't differ.

**Comment by atanasster** [comment]
Yes, i meant if you just want to download/clone the core library without the examples.

I suspect the examples will be significantly larger with image assets etc.

File: grommet/package.json (hunk 1 above)
Comment (ericsoderberghp) [reply grommet/package.json]: Does `jsnext:main` also need changing?

**Comment by ericsoderberghp** [comment]
I'm supportive of the direction.

I am leaning towards moving grommet files down one level to a directory called `grommet`.

I am not too concerned with the examples having large assets. But, we should ensure our examples are using optimized assets, and perhaps even sharing them, or using public ones that don't need to be included in the source.

**Comment by ShimiSun** [comment]
1. need to address:
```
pre-commit: 
pre-commit: We have nothing pre-commit hooks to run. Either you're missing the `scripts`
pre-commit: in your `package.json` or have configured pre-commit to run nothing.
pre-commit: Skipping the pre-commit hook.
pre-commit: 
```
due to directory changes.

2. Adding snapshots to templates

**Comment by ShimiSun** [comment]
allow netlify deployment for the templates

**Comment by ShimiSun** [comment]
Needs attention to recent conflicts that are impacting project structuring and content, such as:
1. Refactoring to hooks
2. Storybook restructuring
3. package.json updates with massive impacts.

Next steps of this PR:
1. Resolve conflicts
2. Allow Netlify deployment for the templates
3. Adding snapshots to templates
4. Add the designer and themer repos

**Comment by ShimiSun** [comment]
All conflicts resolved!
Next topic: 
Go with CircleCI per project https://discuss.circleci.com/t/does-circleci-2-0-work-with-monorepos/10378?gclid=CjwKCAjwtuLrBRAlEiwAPVcZBja34I6iULxdFBgMYYqveTPpI1UbqBOg-Y70ooxKVYyJvLtGJWstCRoCDTsQAvD_BwE

Or use github Actions instead.

**Comment by oorestisime** [comment]
@ShimiSun lets close this no?

**Comment by ShimiSun** [comment]
@oorestisime this PR is my constant reminder for monorepo of grommet+ grommet-site :(

**Pull request closed without merging**